### PR TITLE
Persistent introspection data poc

### PR DESCRIPTION
**Pull request by teskje**
This is a poc for testing the feasibility of our proposed solution for [Introspection for Replicated Compute](https://docs.google.com/document/d/1DBrpKaws7lh8UxJaqsZMFzN-mfQQGOLKS_QpBTyAB_8/edit#heading=h.rlyfnwuds957). The code is quite messy, I have no intention to merge this!

## Example

The poc shows how you can query introspection data produced by any replica on any cluster:

* Create clusters to introspect:
  ```sql
  CREATE CLUSTER c1 REPLICA r1 (SIZE '1'), REPLICA r2 (SIZE '2');
  CREATE CLUSTER c2 REPLICA r1 (SIZE '1');
  ```

* Create an introspection view:
  ```sql
  CREATE MATERIALIZED VIEW mz_dataflow_operators_persist AS (
    SELECT 'c1' AS cluster, 'r1' AS replica, *
    FROM mz_dataflow_operators_2
  ) UNION (
    SELECT 'c1' AS cluster, 'r2' AS replica, *
    FROM mz_dataflow_operators_3
  ) UNION (
    SELECT 'c2' AS cluster, 'r1' AS replica, *
    FROM mz_dataflow_operators_4
  );
  ```
  (A real implementation would create this view automatically.)

* Create some dataflows:
  ```sql
  CREATE TABLE t (foo TEXT);

  SET CLUSTER = c1;
  CREATE MATERIALIZED VIEW v1 AS SELECT COUNT(*) FROM t;

  SET CLUSTER = c2;
  CREATE MATERIALIZED VIEW v2 AS SELECT * FROM t;
  ```

* Query introspection data:
  ```sql
  SET CLUSTER = default;
  SELECT * FROM mz_dataflow_operators_persist;

  SELECT * FROM mz_dataflow_operators_persist WHERE cluster = 'c1' ORDER BY replica;
  ```

## Missing Functionality

* Automatic creation and maintenance of introspection views
* Persisting collection information (shard IDs) in stash
* Loading per-replica introspection catalog state after materialized restart
* Cleanup of introspection sources when replicas are dropped

### CODE REVIEW

File: src/compute/src/logging/persist.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        .expect("could not open persist shard");
+        (Some(write), Some(read))
+    } else {
+        (None, None)
+    };
+
+    // Ensure the collection contains no 0 diffs, as persist chokes on those.
+    // Aside from that, it seems to be a good idea to reduce the data as much
+    // as possible before sending it over the network.
+    let collection = collection.consolidate();
```
Comment (antiguru) [inline src/compute/src/logging/persist.rs:50]: consolidating is very expensive. It's better to use `consolidate_stream`, or make it a requirement to the caller that there are no 0 diffs.

**Review by antiguru — COMMENTED** [review]
This looks like a great start! I'm curious to see what performance implications we get from passing data through persist.

**Comment by teskje** [comment]
Took me way to long to figure out that `persist`'s implementation of file-based storage is blocking! To get any useful benchmark data, S3 must be used. That means changing the hardcoded `blob_uri` to `"s3://<bucket-name>/<bucket-prefix>"` and making sure the bucket is accessible for materialized.

If run benchmarks with `dbbench` on an EC2 r5a.4xlarge instance with a single computed worker and this config:
```ini
duration = 120s

[setup]
query=create table t(a int, b int)
query=create table r(x int, y int)

[teardown]
query=drop table t
query=drop table r

[loadtest]
query=select count(y) from t, r where a = x
rate=300
```

The results look fine. Latency seems to get a bit worse, but not too bad:

- main:
  ```
  2022/05/09 16:38:21 loadtest: 36000 transactions (299.880 TPS), latency 44.430787ms±280.8µs; 36000 rows (299.880 RPS), 36000 queries (299.880 QPS); 0 aborts (0.000%), latency 0s±0s
  Transactions:
    4.194304ms -   8.388608ms [   185]: ▌
    8.388608ms -  16.777216ms [  2946]: ████████▏
   16.777216ms -  33.554432ms [  9523]: ██████████████████████████▍
   33.554432ms -  67.108864ms [ 18002]: ██████████████████████████████████████████████████
   67.108864ms - 134.217728ms [  5339]: ██████████████▊
  134.217728ms - 268.435456ms [     5]: ▏
  ```

- this branch:
  ```
  2022/05/09 16:30:12 loadtest: 36000 transactions (299.798 TPS), latency 47.941843ms±301.045µs; 36000 rows (299.798 RPS), 36000 queries (299.798 QPS); 0 aborts (0.000%), latency 0s±0s
  Transactions:
    4.194304ms -   8.388608ms [    84]: ▏
    8.388608ms -  16.777216ms [  1961]: █████▋
   16.777216ms -  33.554432ms [  9165]: ██████████████████████████▌
   33.554432ms -  67.108864ms [ 17269]: ██████████████████████████████████████████████████
   67.108864ms - 134.217728ms [  7516]: █████████████████████▊
  134.217728ms - 268.435456ms [     5]: ▏
  ```

**Comment by benesch** [comment]
@sploiselle you can hit the subscribe button in the right sidebar to follow without having to comment!

image

**Comment by danhhz** [comment]
> > Took me way to long to figure out that persist's implementation of file-based storage is blocking!
> 
> Doh! I forgot about that. Long story. PR incoming

Sorry it took a sec, a minor perf investigation was required but this is now merging on green. https://github.com/MaterializeInc/materialize/pull/12360

**Comment by teskje** [comment]
Thanks Dan! It's really my fault trying to use a dev feature for benchmarking :)

**Pull request closed without merging**